src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,55 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private T[] elements;
+    private int size;
+    private static final int INITIAL_CAPACITY = 10;
+
+    @SuppressWarnings("unchecked")
+    public MateStack() {
+        elements = (T[]) new Object[INITIAL_CAPACITY];
+        size = 0;
+    }
 
+    public void push(T value) {
+        ensureCapacity();
+        elements[size++] = value;
     }
 
     public T peek() {
-        return null;
+        ensureNotEmpty();
+        return elements[size - 1];
     }
 
     public T pop() {
-        return null;
+        ensureNotEmpty();
+        T value = elements[--size];
+        elements[size] = null;
+        return value;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void resize() {
+        T[] newArray = (T[]) new Object[elements.length * 2];
+        System.arraycopy(elements, 0, newArray, 0, elements.length);
+        elements = newArray;
+    }
```

Comment:
In the `resize()` method, you are doubling the size of the array each time it is full. While this is a common approach, it can lead to a lot of wasted space if the stack size decreases. You might want to consider adding logic to decrease the size of the array when the stack size decreases to a certain point. This can help to save memory.